File: packages/android-kotlin/src/main/java/com/intosoft/qrcode/IntosoftQRCode.kt
package com.intosoft.qrcode

import android.content.Context
import com.eclipsesource.v8.V8
import com.eclipsesource.v8.V8Object
import com.google.gson.Gson
import java.io.BufferedReader
import java.io.InputStreamReader

enum class BodyShape(val value: String) {
    DOTS("dots"),
    CLASSY("classy")
}

enum class EyeballShape(val value: String) {
    LEAF("leaf"),
    POINTED("pointed"),
    EXTRA_ROUNDED("extra-rounded")
}

enum class EyeFrameShape(val value: String) {
    LEAF("leaf"),
    POINTED("pointed")
}

data class GradientConfig(
    val type: String,
    val colors: List<String>,
    val rotation: Double? = null
)

data class LogoConfig(
    val url: String,
    val size: Double = 0.15,
    val removeBackground: Boolean = false
)

data class QRCodeConfig(
    val text: String,
    val width: Int = 200,
    val height: Int = 200,
    val margin: Int = 10,
    val dotsColor: String = "#000000",
    val backgroundColor: String = "#ffffff",
    val bodyShape: BodyShape? = null,
    val eyeballShape: EyeballShape? = null,
    val eyeFrameShape: EyeFrameShape? = null,
    val gradient: GradientConfig? = null,
    val logo: LogoConfig? = null
)

class IntosoftQRCode private constructor() {
    companion object {
        private var v8Runtime: V8? = null
        private var isInitialized = false
        private val gson = Gson()

        @Synchronized
        private fun initialize(context: Context) {
            if (isInitialized) return

            v8Runtime = V8.createV8Runtime()
            
            val jsCode = context.assets.open("qrcode.js").use { inputStream ->
                BufferedReader(InputStreamReader(inputStream)).use { reader ->
                    reader.readText()
                }
            }
            
            v8Runtime?.executeScript(jsCode)
            isInitialized = true
        }

        fun generateSVG(context: Context, config: QRCodeConfig): String {
            initialize(context)
            
            val runtime = v8Runtime ?: throw QRCodeException("V8 runtime not initialized")
            
            val configMap = mutableMapOf<String, Any>(
                "text" to config.text,
                "width" to config.width,
                "height" to config.height,
                "margin" to config.margin,
                "dotsColor" to config.dotsColor,
                "bgColor" to config.backgroundColor
            )
            
            config.bodyShape?.let { configMap["bodyShape"] = it.value }
            config.eyeballShape?.let { configMap["eyeballShape"] = it.value }
            config.eyeFrameShape?.let { configMap["eyeFrameShape"] = it.value }
            
            config.gradient?.let { gradient ->
                configMap["gradient"] = mapOf(
                    "type" to gradient.type,
                    "colors" to gradient.colors,
                    "rotation" to gradient.rotation
                ).filterValues { it != null }
            }
            
            config.logo?.let { logo ->
                configMap["logo"] = mapOf(
                    "url" to logo.url,
                    "size" to logo.size,
                    "removeBackground" to logo.removeBackground
                )
            }
            
            val configJson = gson.toJson(configMap)
            
            val script = """
                (function() {
                    try {
                        const config = $configJson;
                        return window.qrcode.generateSVGString(config);
                    } catch (error) {
                        return JSON.stringify({ error: error.message });
                    }
                })()
            """.trimIndent()
            
            val result = runtime.executeStringScript(script)
            
            if (result.contains("\"error\"")) {
                throw QRCodeException("JavaScript error: $result")
            }
            
            return result
        }

        fun generateSVG(
            context: Context,
            text: String,
            width: Int = 200,
            height: Int = 200,
            dotsColor: String = "#000000",
            backgroundColor: String = "#ffffff"
        ): String {
            val config = QRCodeConfig(
                text = text,
                width = width,
                height = height,
                dotsColor = dotsColor,
                backgroundColor = backgroundColor
            )
            return generateSVG(context, config)
        }

        fun release() {
            v8Runtime?.release()
            v8Runtime = null
            isInitialized = false
        }
    }
}

class QRCodeException(message: String) : Exception(message)

File: packages/ios-swift/Sources/IntosoftQRCode/IntosoftQRCode.swift
import Foundation
import JavaScriptCore

public enum BodyShape: String {
    case dots
    case classy
}

public enum EyeballShape: String {
    case leaf
    case pointed
    case extraRounded = "extra-rounded"
}

public enum EyeFrameShape: String {
    case leaf
    case pointed
}

public struct GradientConfig {
    public let type: String
    public let colors: [String]
    public let rotation: Double?
    
    public init(type: String, colors: [String], rotation: Double? = nil) {
        self.type = type
        self.colors = colors
        self.rotation = rotation
    }
}

public struct LogoConfig {
    public let url: String
    public let size: Double
    public let removeBackground: Bool
    
    public init(url: String, size: Double = 0.15, removeBackground: Bool = false) {
        self.url = url
        self.size = size
        self.removeBackground = removeBackground
    }
}

public struct QRCodeConfig {
    public var text: String
    public var width: Int
    public var height: Int
    public var margin: Int
    public var dotsColor: String
    public var backgroundColor: String
    public var bodyShape: BodyShape?
    public var eyeballShape: EyeballShape?
    public var eyeFrameShape: EyeFrameShape?
    public var gradient: GradientConfig?
    public var logo: LogoConfig?
    
    public init(
        text: String,
        width: Int = 200,
        height: Int = 200,
        margin: Int = 10,
        dotsColor: String = "#000000",
        backgroundColor: String = "#ffffff",
        bodyShape: BodyShape? = nil,
        eyeballShape: EyeballShape? = nil,
        eyeFrameShape: EyeFrameShape? = nil,
        gradient: GradientConfig? = nil,
        logo: LogoConfig? = nil
    ) {
        self.text = text
        self.width = width
        self.height = height
        self.margin = margin
        self.dotsColor = dotsColor
        self.backgroundColor = backgroundColor
        self.bodyShape = bodyShape
        self.eyeballShape = eyeballShape
        self.eyeFrameShape = eyeFrameShape
        self.gradient = gradient
        self.logo = logo
    }
}

public class IntosoftQRCode {
    private static var jsContext: JSContext?
    private static var isInitialized = false
    
    private static func initialize() throws {
        guard !isInitialized else { return }
        
        guard let jsPath = Bundle.module.url(forResource: "qrcode", withExtension: "js", subdirectory: "Resources"),
              let jsCode = try? String(contentsOf: jsPath) else {
            throw QRCodeError.failedToLoadJSBundle
        }
        
        jsContext = JSContext()
        jsContext?.exceptionHandler = { context, exception in
            print("JS Error: \(exception?.toString() ?? "unknown error")")
        }
        
        jsContext?.evaluateScript(jsCode)
        isInitialized = true
    }
    
    public static func generateSVG(config: QRCodeConfig) throws -> String {
        try initialize()
        
        guard let context = jsContext else {
            throw QRCodeError.jsContextNotInitialized
        }
        
        var configDict: [String: Any] = [
            "text": config.text,
            "width": config.width,
            "height": config.height,
            "margin": config.margin,
            "dotsColor": config.dotsColor,
            "bgColor": config.backgroundColor
        ]
        
        if let bodyShape = config.bodyShape {
            configDict["bodyShape"] = bodyShape.rawValue
        }
        
        if let eyeballShape = config.eyeballShape {
            configDict["eyeballShape"] = eyeballShape.rawValue
        }
        
        if let eyeFrameShape = config.eyeFrameShape {
            configDict["eyeFrameShape"] = eyeFrameShape.rawValue
        }
        
        if let gradient = config.gradient {
            var gradientDict: [String: Any] = [
                "type": gradient.type,
                "colors": gradient.colors
            ]
            if let rotation = gradient.rotation {
                gradientDict["rotation"] = rotation
            }
            configDict["gradient"] = gradientDict
        }
        
        if let logo = config.logo {
            configDict["logo"] = [
                "url": logo.url,
                "size": logo.size,
                "removeBackground": logo.removeBackground
            ]
        }
        
        guard let configData = try? JSONSerialization.data(withJSONObject: configDict),
              let configJSON = String(data: configData, encoding: .utf8) else {
            throw QRCodeError.invalidConfiguration
        }
        
        let script = """
        (function() {
            try {
                const config = \(configJSON);
                return window.qrcode.generateSVGString(config);
            } catch (error) {
                return JSON.stringify({ error: error.message });
            }
        })()
        """
        
        guard let result = context.evaluateScript(script),
              let svgString = result.toString(),
              !svgString.isEmpty else {
            throw QRCodeError.failedToGenerateSVG
        }
        
        if svgString.contains("\"error\"") {
            throw QRCodeError.jsError(svgString)
        }
        
        return svgString
    }
    
    public static func generateSVG(
        text: String,
        width: Int = 200,
        height: Int = 200,
        dotsColor: String = "#000000",
        backgroundColor: String = "#ffffff"
    ) throws -> String {
        let config = QRCodeConfig(
            text: text,
            width: width,
            height: height,
            dotsColor: dotsColor,
            backgroundColor: backgroundColor
        )
        return try generateSVG(config: config)
    }
}

public enum QRCodeError: Error {
    case failedToLoadJSBundle
    case jsContextNotInitialized
    case invalidConfiguration
    case failedToGenerateSVG
    case jsError(String)
}

extension QRCodeError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .failedToLoadJSBundle:
            return "Failed to load QR code JavaScript bundle"
        case .jsContextNotInitialized:
            return "JavaScript context not initialized"
        case .invalidConfiguration:
            return "Invalid QR code configuration"
        case .failedToGenerateSVG:
            return "Failed to generate SVG string"
        case .jsError(let message):
            return "JavaScript error: \(message)"
        }
    }
}
